### PR TITLE
[serve] Refactor router policies to remove inheritance

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Currently different policies are implemented as subclasses of the Router. This is a bit ugly/hard to follow in the code because the subclass overrides one (and sometimes two) of the methods of the router, and also directly modifies its fields. Additionally, we want to be able to specify a different policy per endpoint (and potentially per backend) in the future.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/serve/policy.py [hunk 1]
```diff
@@ -3,43 +3,37 @@
 
 import numpy as np
 
-import ray
-from ray.serve.router import Router
 from ray.serve.utils import logger
```
Comment (simon-mo) [inline python/ray/serve/policy.py:7]: define a base class with `async def flush(self, endpoint_queue_backend_queues) as signature?

File: python/ray/serve/policy.py [hunk 2]
```diff
@@ -3,43 +3,37 @@
...
-            # while there are incoming requests and there are backends
-            while queue.qsize() and len(self.traffic[endpoint]):
-                backend_names = list(self.traffic[endpoint].keys())
-                backend_weights = list(self.traffic[endpoint].values())
-                # randomly choose a backend for every query
-                chosen_backend = np.random.choice(
-                    backend_names, replace=False, p=backend_weights).squeeze()
-                logger.debug("Matching endpoint {} to backend {}".format(
-                    endpoint, chosen_backend))
+    def __init__(self, endpoint, traffic_dict):
```
Comment (simon-mo) [inline python/ray/serve/policy.py:15]: do we really need endpoint as constructor argument here? in router the `self.traffic` dict is already keyed by endpoint.

**Review by simon-mo — COMMENTED** [review]
yeah the direction looks good to me, since we are just changing we policies are passed to router.

File: python/ray/serve/policy.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/policy.py:15]: We don't, but it's convenient for logging.

File: python/ray/serve/policy.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/policy.py:7]: will do

File: python/ray/serve/policy.py [hunk 3]
```diff
@@ -1,188 +1,167 @@
...
-    `Stateless` meaning the current decisions of selecting backend are
-    not dependent on previous decisions. Random policy (randomly) samples
-    backends based on backend weights for every query. This policy uses the
-    weights assigned to backends.
+class RoutingPolicy:
+    """Defines the interface for a routing policy for a single endpoint.
+
+    To add a new routing policy, a class should be defined that provides this
+    interface. The class may be stateful, in which case it may also want to
+    provide a non-default constructor.
```
Comment (simon-mo) [inline python/ray/serve/policy.py:15]: Stateful but not the state will not be restored on router reconstruct.

File: python/ray/serve/router.py [hunk 4]
```diff
@@ -261,14 +267,16 @@ async def remove_worker(self, backend_tag, replica_tag):
     async def set_traffic(self, endpoint, traffic_dict):
         logger.debug("Setting traffic for endpoint %s to %s", endpoint,
                      traffic_dict)
-        self.traffic[endpoint] = traffic_dict
-        await self.flush()
+        self.traffic[endpoint] = self.policy(traffic_dict,
+                                             **self.policy_kwargs)
+
```
Comment (simon-mo) [inline python/ray/serve/router.py:272]: I think we need to grab the flush lock here? There could a race condition where we in the middle of flushing endpoint queue somewhere and the policy got replaced. So we end up flushing to old backends. But I guess this is fine since we are grabbing the lock immediately afterward.

Comment (simon-mo) [reply python/ray/serve/router.py:272]: ok i think the issue occurs when other coroutines _with_ the flush lock will inspect the `self.traffic` and assume a consistent view, but here we can mutate it _without_ the lock.

**Review by simon-mo — APPROVED** [review]

File: python/ray/serve/policy.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/policy.py:15]: Good point

File: python/ray/serve/router.py (hunk 4 above)
Comment (edoakes) [reply python/ray/serve/router.py:272]: Good catch, put it under the lock.
